casper/src/test/scala/coop/rchain/casper/api/LastFinalizedAPITest.scala

```diff
@@ -0,0 +1,145 @@
+    validatorKeyPairs,
+    Map(validatorPks(0) -> 10, validatorPks(1) -> 10, validatorPks(2) -> 5)
+  )
+  val genesisContext = buildGenesis(genesisParameters)
+
+  implicit val metricsEff = new Metrics.MetricsNOP[Task]
+
+  def isFinalized(block: BlockMessage)(engineCell: Cell[Task, Engine[Task]])(
+      implicit blockStore: BlockStore[Task],
+      safetyOracle: SafetyOracle[Task],
+      log: Log[Task]
+  ): Task[Boolean] =
+    BlockAPI
+      .isFinalized[Task](
+        IsFinalizedQuery(hash = ProtoUtil.hashString(block))
+      )(Sync[Task], engineCell, safetyOracle, blockStore, log)
+      .map(
+        _.right.value.isFinalized
+      )
+
+  /*
+   * DAG Looks like this:
+   *
+   *           b7
+   *           |
+   *           b6
+   *           |
+   *           b5 <- last finalized block
+   *         / |
+   *        |  b4
+   *        |  |
+   *       b2  b3
+   *         \ |
+   *           b1
+   *           |
+   *         genesis
+   */
+  "isFinalized" should "return true for ancestors of last finalized block" in effectTest {
+    HashSetCasperTestNode.networkEff(genesisContext, networkSize = 3).use {
+      case n1 +: n2 +: n3 +: _ =>
+        import n1.{blockStore, cliqueOracleEffect, logEff}
+        val engine = new EngineWithCasper[Task](n1.casperEff)
+        val nodes  = List(n1, n2, n3)
+        for {
+          produceDeploys <- (0 until 7).toList.traverse(i => basicDeployData[Task](i))
+
+          b1 <- n1.publishBlock(produceDeploys(0))(nodes: _*)
+          b2 <- n2.publishBlock(produceDeploys(1))()
+          b3 <- n3.publishBlock(produceDeploys(2))(n1)
+          b4 <- n1.publishBlock(produceDeploys(3))(nodes: _*)
+          b5 <- n2.publishBlock(produceDeploys(4))(nodes: _*)
+          b6 <- n1.publishBlock(produceDeploys(5))(nodes: _*)
+          b7 <- n2.publishBlock(produceDeploys(6))(nodes: _*)
+
+          lastFinalizedBlock <- n1.casperEff.lastFinalizedBlock
+          _                  = lastFinalizedBlock shouldBe b5
+
+          engineCell <- Cell.mvarCell[Task, Engine[Task]](engine)
+          // Checking if last finalized block is finalized
+          _ <- isFinalized(b5)(engineCell) shouldBeF true
+          // Checking if parent of last finalized block is finalized
+          _ <- isFinalized(b4)(engineCell) shouldBeF true
+          // Checking if secondary parent of last finalized block is finalized
+          _ <- isFinalized(b2)(engineCell) shouldBeF true
+        } yield ()
+    }
+  }
+
+  /*
+   * DAG Looks like this:
+   *
+   *           b5
+   *           |
+   *           b4
+   *           |
+   *       b2  b3 <- last finalized block
+   *         \ |
+   *           b1
+   *           |
+   *         genesis
+   */
+  it should "return false for children and siblings of last finalized block" in effectTest {
+    HashSetCasperTestNode.networkEff(genesisContext, networkSize = 3).use {
+      case n1 +: n2 +: n3 +: _ =>
+        import n3.{blockStore, cliqueOracleEffect, logEff}
```

Comment:
why do we differ the 'focused' node between those tests?

I have a hard time understanding which node sees what based on the picture. Maybe each node should have its own swim lane?

Reply:
No particular reason. I changed so that we always look from the n1 perspective.

Not sure what exactly do you mean by "swim lane" here but describing the whole state from all nodes POV would be overly verbose. Note that there is also "time" axis involved here: a node can get a block later than the other nodes (e.g. by missing it initially but fetching it as a dependency for a new one) which can affect node's decision on finalization. This is what happens in the second test (`it should "return false for children and siblings of last finalized block"`).